expression/builtin_cast.go

```diff
@@ -1983,3 +1986,87 @@ func WrapWithCastAsJSON(ctx sessionctx.Context, expr Expression) Expression {
 	}
 	return BuildCastFunction(ctx, expr, tp)
 }
+
+// TryPushCastDownToControlFunctionForHybridType try to push cast down to control function for Hybrid Type.
+// If necessary, it will rebuild control function using changed args.
+// When a hybrid type is the output of a control function, the result may be as a numeric type to subsequent calculation
+// We should perform the `Cast` operation early to avoid using the wrong type for calculation
+func TryPushCastDownToControlFunctionForHybridType(ctx sessionctx.Context, expr Expression, tp *types.FieldType) (res Expression, success bool) {
+	sf, ok := expr.(*ScalarFunction)
+	if !ok {
+		return expr, false
+	}
+
+	var wrapCastFunc func(ctx sessionctx.Context, expr Expression) Expression
+	switch tp.EvalType() {
+	case types.ETInt:
+		wrapCastFunc = WrapWithCastAsInt
+	case types.ETReal:
+		wrapCastFunc = WrapWithCastAsReal
+	case types.ETString:
+		wrapCastFunc = WrapWithCastAsString
+	default:
+		return expr, false
+	}
+
+	args := sf.GetArgs()
+	switch sf.FuncName.L {
+	case ast.If:
+		if args[1].GetType().Hybrid() || args[2].GetType().Hybrid() {
+			args[1] = wrapCastFunc(ctx, args[1])
+			args[2] = wrapCastFunc(ctx, args[2])
+			f, err := funcs[ast.If].getFunction(ctx, args)
+			if err != nil {
+				return expr, false
+			}
+			sf.RetType, sf.Function = f.getRetTp(), f
+			return sf, true
+		}
+	case ast.Case:
+		hasHybrid := false
+		for i := 0; i < len(args)-1; i += 2 {
+			hasHybrid = hasHybrid || args[i+1].GetType().Hybrid()
+		}
+		if len(args)%2 == 1 {
+			hasHybrid = hasHybrid || args[len(args)-1].GetType().Hybrid()
+		}
+		if !hasHybrid {
+			return expr, false
+		}
+
+		for i := 0; i < len(args)-1; i += 2 {
+			args[i+1] = wrapCastFunc(ctx, args[i+1])
+		}
+		if len(args)%2 == 1 {
+			args[len(args)-1] = wrapCastFunc(ctx, args[len(args)-1])
+		}
+		f, err := funcs[ast.Case].getFunction(ctx, args)
+		if err != nil {
+			return expr, false
+		}
+		sf.RetType, sf.Function = f.getRetTp(), f
+		return sf, true
+	case ast.Elt:
+		hasHybrid := false
+		for i := 1; i < len(args); i++ {
+			hasHybrid = hasHybrid || args[i].GetType().Hybrid()
+		}
+		if !hasHybrid {
+			return expr, false
+		}
+
+		for i := 1; i < len(args); i++ {
+			args[i] = wrapCastFunc(ctx, args[i])
+		}
+		f, err := funcs[ast.Elt].getFunction(ctx, args)
+		if err != nil {
+			return expr, false
+		}
+		sf.RetType, sf.Function = f.getRetTp(), f
+		// Elt only supports ETString, so we need add extra cast to keep retType right.
+		return wrapCastFunc(ctx, sf), true
```

Comment:
The invoker will wrap a cast, why do we need to do this here?

Reply:
Ok, I remove this.